python/kwiver/vital/tests/test_activity_type.py

```diff
@@ -0,0 +1,95 @@
+
+ * Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+ * Neither name of Kitware, Inc. nor the names of any contributors may be used
+   to endorse or promote products derived from this software without specific
+   prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS ``AS IS''
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+==============================================================================
+
+Tests for ActivityType interface class.
+
+"""
+import unittest
+import nose.tools as nt
+import numpy as np
+
+
+from kwiver.vital.types import ActivityType as AT
+
+
+class TestActivityType(unittest.TestCase):
+
+    def test_constructor(self):
+        AT()
+        AT(np.array(["name1","class_name2","class3"]),np.array([1.0,2.3,3.14]))
+        AT("name", 2.0)
+
+    def test_methods(self):
+        t = AT(np.array(["name1","class_name2","class3"]),np.array([1.0,2.3,3.14]))
+
+        # str/repr/itr
+        self.assertIsInstance(str(t),str)
+        self.assertIsInstance(t.__repr__(), str)
+        itr = iter(t)
+        self.assertIsInstance(next(itr),tuple)
+
+        # Has Class Name
+        self.assertTrue(t.has_class_name("name1"))
+        self.assertFalse(t.has_class_name("foo"))
+
+        # Score
+        self.assertEqual(t.score("class_name2"),2.3)
+
+        # Get most likely
+        self.assertEqual(t.get_most_likely_class(),"class3")
+        self.assertEqual(3.14,t.get_most_likely_score())
+
+        # Set Score
+        t.set_score("foo1",3.8)
+        self.assertEqual(3.8,t.score("foo1"))
+        t.set_score("foo2",3.9)
+        self.assertTrue(t.has_class_name("foo2"))
+        self.assertEqual(t.score("foo2"),3.9)
+
+        # Delete Score
+        t.delete_score("foo1")
+        self.assertFalse(t.has_class_name("foo1"))
+
+        # Class Names
+        np.testing.assert_array_equal(t.class_names(), np.array(["foo2","class3","class_name2","name1"]))
+        np.testing.assert_array_equal(t.class_names(3.0), np.array(["foo2","class3"]))
+
+        np.testing.assert_array_equal(t.all_class_names(),np.array(["class3","class_name2","foo1","foo2","name","name1"]))
```

Comment:
Should we always get the same order here? [This](http://www.cplusplus.com/reference/unordered_set/unordered_set/begin/#:~:text=std%3A%3Aunordered_set%3A%3Abegin&text=Returns%20an%20iterator%20pointing%20to,is%20considered%20its%20first%20element.) seems to say otherwise. Same goes for the detected_object_set_type tests

Reply:
Ugh. No, this should not be using an order-dependent comparison. Come to think of it, I wonder if `all_class_names` can be changed to return an `unordered_set`? (Then it would be obvious that no order is expected.) IIRC we used to use `set`, which would have been in lexicographical order...

Reply:
...and also the doc of `all_class_names` says it's supposed to be sorted. What do you (and/or @linus-sherrill, @dstoup, or anyone else that want to chime in) think? Should we revert to `std::set` or change `all_class_names` to sort the output? (Do we worry about optimizing for look-ups against a larger `s_master_name_set` and assume that `all_class_names` isn't called in performance-critical code?)

Reply:
...or do we just return an unordered_set (I somewhat wonder anyway about not just returning the same container type used internally) and punt this to users?

Reply:
Sorted `all_class_names` was an artifact of the implementation, but it is in the documentation so it may be relied upon. I think is is reasonable for that method to sort the names before returning them, since that is not a common call.

Reply:
That works for me